### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by quarckster**

### CODE REVIEW

**Comment by rsnyman** [comment]
This looks good. Nice separation of models and logic.

What are your thoughts on renaming `pytest_ibutsu.py` to something like `pytest_plugin.py` ?

File: src/pytest_ibutsu/modeling.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    def increment(self, test_result: "TestResult") -> None:
+        attr = {
+            "failed": "failures",
+            "error": "errors",
+            "skipped": "skips",
+            "xfailed": "xfailures",
+            "xpassed": "xpasses",
+        }.get(test_result.result)
+        if attr:
+            inc = getattr(self, attr)
```
[the remark this answers is not in the document]
Comment (quarckster) [reply src/pytest_ibutsu/modeling.py:53]: sure, thanks Mike for the review

File: src/pytest_ibutsu/modeling.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+            setattr(self, attr, inc + 1)
+        self.tests += 1
+
+
+@define
+class TestRun:
+    component: Optional[str] = None
+    env: Optional[str] = None
+    id: str = field(factory=lambda: str(uuid.uuid4()))
```
Comment (mshriver) [inline src/pytest_ibutsu/modeling.py:62]: Not finding any instantiations of this class in the PR, would `id` be passed ever? (`init=False` kwarg for field)

File: tests/test_archive.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    del result["metadata"]["durations"]
+    return result
+
+
+@pytest.fixture
+def run_id():
+    return str(uuid.uuid4())
+
+
+@pytest.fixture(params=PYTEST_ARGS, ids=["no-xdist", "xdist"])
```
Comment (RonnyPfannschmidt) [inline tests/test_archive.py:31]: use `pytest.param` here to ensure co-location of param data and test ids here,

File: src/pytest_ibutsu/modeling.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+                return item.name
+
+    @classmethod
+    def from_item(cls, item: pytest.Item) -> "TestResult":
+        return cls(
+            test_id=cls._get_test_idents(item),
+            params=cls._get_item_params(item),
+            source=item.config.ibutsu_plugin.ibutsu_source,
+            run_id=item.config.ibutsu_plugin.run.id,
+            metadata={
```
[the remark this answers is not in the document]
Comment (mshriver) [reply src/pytest_ibutsu/modeling.py:207]: AH! I see now that they're getting pulled in later during `pytest_runtest_logreport` from the report object.

File: src/pytest_ibutsu/modeling.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+        except AttributeError:
+            try:
+                return item.fspath.strpath
+            except AttributeError:
+                return item.name
+
+    @classmethod
+    def from_item(cls, item: pytest.Item) -> "TestResult":
+        return cls(
+            test_id=cls._get_test_idents(item),
```
[the remark this answers is not in the document]
Comment (quarckster) [reply src/pytest_ibutsu/modeling.py:203]: `test_id` is defined in Ibutsu server OpenAPI spec https://github.com/ibutsu/ibutsu-server/blob/master/backend/ibutsu_server/openapi/openapi.yaml#L2182.

[the remark this answers is not in the document]
Comment (mshriver) [reply src/pytest_ibutsu/modeling.py:203]: ack, but its insufficient to be unique as described. 

I think we need a separate `nodeid` field added, or the existing `id` expanded to include more that the test function name. 

Doesn't have to be part of this PR.

[the remark this answers is not in the document]
Comment (quarckster) [reply src/pytest_ibutsu/modeling.py:203]: I agree that the current values in `test_id` field are not unique enough. But I wouldn't want to break the existing data structure. Let's make this change in the next release. Please create an issue for it. It's definitely worth implementing. As for `id` field, it contains a UUID string and we can consider it as unique.

**Review by rsnyman — APPROVED** [review]
This all looks good to me. Presuming everyone else is happy, and this is still compatible with IQE, then I'm happy to merge.

File: src/pytest_ibutsu/modeling.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mshriver) [reply src/pytest_ibutsu/modeling.py:203]: They're unique, but to an exact test result, not the case itself.

I don't want to go off the deep end here (polarion) and try to manage 'test cases' per say, but at least give someone the ability to see test history for the same test function in the same test module, with the same parameter.

**Review by rsnyman — APPROVED** [review]

**Pull request merged**